server/src/main/java/org/candlepin/dto/SimpleModelTranslator.java

```diff
@@ -33,63 +35,171 @@
+            throw new IllegalArgumentException("outputClass is null");
+        }
+
+        Map<Class, ObjectTranslator> inputMappings = this.translators.get(outputClass);
+        if (inputMappings == null) {
+            inputMappings = new HashMap<Class, ObjectTranslator>();
+            this.translators.put(outputClass, inputMappings);
+        }
+
+        ObjectTranslator<I, O> existing = (ObjectTranslator<I, O>) inputMappings.get(inputClass);
+        inputMappings.put(inputClass, translator);
 
         return existing;
     }
 
     /**
      * {@inheritDoc}
      */
-    public ObjectTranslator unregisterTranslator(Class srcClass) {
-        if (srcClass == null) {
-            throw new IllegalArgumentException("srcClass is null");
+    @Override
+    public <I, O> ObjectTranslator<I, O> unregisterTranslator(Class<I> inputClass, Class<O> outputClass) {
+        if (inputClass == null) {
+            throw new IllegalArgumentException("inputClass is null");
+        }
+
+        if (outputClass == null) {
+            throw new IllegalArgumentException("outputClass is null");
+        }
+
+        Map<Class, ObjectTranslator> inputMappings = this.translators.get(outputClass);
+        return inputMappings != null ? (ObjectTranslator<I, O>) inputMappings.remove(inputClass) : null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int unregisterTranslator(ObjectTranslator translator) {
+        if (translator == null) {
+            throw new IllegalArgumentException("translator is null");
+        }
+
+        int mappings = 0;
+
+        for (Map<Class, ObjectTranslator> inputMappings : this.translators.values()) {
+            Iterator<ObjectTranslator> translators = inputMappings.values().iterator();
+            while (translators.hasNext()) {
+                ObjectTranslator existing = translators.next();
+
+                if (translator.equals(existing)) {
+                    translators.remove();
+                    ++mappings;
+                }
+            }
         }
 
-        return this.translators.remove(srcClass);
+        return mappings;
     }
 
     /**
      * {@inheritDoc}
      */
-    public ObjectTranslator getTranslator(Class srcClass) {
-        if (srcClass == null) {
-            throw new IllegalArgumentException("srcClass is null");
+    public <I, O> ObjectTranslator<I, O> getTranslator(Class<I> inputClass, Class<O> outputClass) {
+        if (inputClass == null) {
+            throw new IllegalArgumentException("inputClass is null");
+        }
+
+        if (outputClass == null) {
+            throw new IllegalArgumentException("outputClass is null");
+        }
+
+        Map<Class, ObjectTranslator> inputMappings = this.translators.get(outputClass);
+        return inputMappings != null ? (ObjectTranslator<I, O>) inputMappings.get(inputClass) : null;
+    }
+
+    /**
+     * Attempts to find the nearest class to the given source class within the set of provided
```

Comment:
What does "nearest" mean?

Reply:
Nearest parent-ish class to the given class, if the class itself is not already mapped.